libarchive/archive_util.c

```diff
@@ -701,3 +714,125 @@ archive_utility_string_sort(char **strings)
 		size++;
 	  return archive_utility_string_sort_helper(strings, size);
 }
+
+// This should be enough to find "cpu cores<whitespace>:<space><number>" at the start of a line in /proc/cpuinfo
+#define PROC_CPUINFO_BUF_SIZE 80
+
+int
+__archive_num_physical_cores()
+{
+	static int cores = 0;
+	if (cores > 0) return cores;
+
+#if defined(__FreeBSD__)
+	int sysctlbynameCores = 0;
+	size_t size = sizeof(sysctlbynameCores);
+	if (sysctlbyname("kern.smp.cores", &sysctlbynameCores, &size, NULL, 0) != 0)
+		goto unixfallback;
+
+	if (sysctlbynameCores <= 0) goto unixfallback;
+	cores = sysctlbynameCores;
+	return cores;
+#endif
+
+#ifdef __APPLE__
+	int32_t sysctlbynameCores = 0;
+	size_t size = sizeof(sysctlbynameCores);
+	if (sysctlbyname("hw.physicalcpu", &sysctlbynameCores, &size, NULL, 0) != 0)
+		goto unixfallback;
+
+	if (sysctlbynameCores <= 0) goto unixfallback;
+	cores = sysctlbynameCores;
+	return cores;
+#endif // __APPLE__
+
+#if defined(__linux__)
+	// Look in /proc/cpuinfo for a line like "cpu cores       : <num>"
+
+	FILE *fp = fopen("/proc/cpuinfo", "r");
```

Comment:
That's strange. Why not to use [`get_nprocs()`](https://linux.die.net/man/3/get_nprocs) for Linux instead of all that shenanigans?

Reply:
`get_nprocs()` returns the number of logical cores, but this feature tries to find and use the number of physical cores.

Reply:
Hmm, are you sure it's necessary? SMT grants a fairly large increase.

Also I believe that the process should respect CPU affinity.
Here is how it's done by xz https://github.com/tukaani-project/xz/blob/master/src/common/tuklib_cpucores.c

Reply:
Actually nevermind. I see now that you are effectively followed [the implementation from `zstd` CLI](https://github.com/facebook/zstd/blob/889392dac8fb88b49d6626783acbeff51d4386f2/programs/util.c#L1476-L1545).
Makes sense to do it that way then.

Reply:
Yeah- I figured it makes most sense to behave like the zstd util.